### PR TITLE
Extend performance contract with custom measurements

### PR DESCRIPTION
**Pull request by dynco-nym**
New APIs
- define measurement
- retire measurement

Existing APIs
- extended so that measurement kind needs to be specified
- all performance metrics now need to be defined under some measurement kind

To do
- more unit testing of edge cases
- handle staleness check before deploying (commented in the code about it)

- - -
This change is [Reviewable](https://reviewable.io/reviews/nymtech/nym/6282)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -34,6 +34,7 @@ members = [
     "common/nym-common",
     "common/config",
     "common/cosmwasm-smart-contracts/coconut-dkg",
+    "contracts/performance",
```
Comment (jstuczyn) [inline Cargo.toml:37]: don't add it to the main workspace - all contracts have their own sub workspace

File: contracts/performance/src/storage.rs [hunk 2]
```diff
@@ -497,6 +559,76 @@ impl PerformanceResultsStorage {
...
+    }
+
+    /// assumes authorization had been performed
+    pub fn define_new_measurement_kind(
+        &self,
+        storage: &mut dyn Storage,
+        measurement_kind: MeasurementKind,
+    ) -> Result<(), NymPerformanceContractError> {
+        if self.is_measurement_defined(storage, measurement_kind.clone())? {
+            return Err(NymPerformanceContractError::InvalidInput(format!(
```
Comment (jstuczyn) [inline contracts/performance/src/storage.rs]: that's a nit, so feel free to ignore it, but I think it would be cleaner if `MeasurementAlreadyDefined` was its own enum variant rather than being put alongside different errors within the opaque `InvalidInput`

File: contracts/performance/src/storage.rs [hunk 3]
```diff
@@ -497,6 +559,76 @@ impl PerformanceResultsStorage {
...
+    }
+
+    /// assumes authorization had been performed
+    pub fn retire_measurement_kind(
+        &self,
+        storage: &mut dyn Storage,
+        measurement_kind: MeasurementKind,
+    ) -> Result<(), NymPerformanceContractError> {
+        if !self.is_measurement_defined(storage, measurement_kind.clone())? {
+            return Err(NymPerformanceContractError::InvalidInput(format!(
```
Comment (jstuczyn) [inline contracts/performance/src/storage.rs]: same nit to do with the error variant

File: contracts/performance/src/transactions.rs [hunk 4]
```diff
@@ -21,6 +21,50 @@ pub fn try_update_contract_admin(
...
+        .contract_admin
+        .assert_admin(deps.as_ref(), sender)?;
+
+    // validation
+    if measurement_kind.len() < 2
+        || measurement_kind.len() > 20
+        || !measurement_kind.is_ascii()
+        || measurement_kind.contains(char::is_whitespace)
+    {
+        return Err(NymPerformanceContractError::InvalidInput(format!(
```
Comment (jstuczyn) [inline contracts/performance/src/transactions.rs]: I think it'd be useful to return error message saying which particular check failed

File: contracts/performance/src/transactions.rs [hunk 5]
```diff
@@ -311,4 +365,200 @@ mod tests {
...
+            },
+        };
+
+        #[allow(clippy::panic)]
+        #[test]
+        fn add_requires_admin() {
+            let mut tester = init_contract_tester();
+            let admin = tester.admin_msg();
+            let new_measurement = MeasurementKind::from("new-measurement");
+
```
Comment (jstuczyn) [inline contracts/performance/src/transactions.rs:401]: nit-ish. technically do validate it does indeed require admin privileges, you have to check that it does error out from any other account

File: contracts/performance/src/transactions.rs [hunk 6]
```diff
@@ -311,4 +365,200 @@ mod tests {
...
+            let first_attempt = try_define_measurement_kind(
+                tester.deps_mut(),
+                &admin.sender,
+                new_measurement.clone(),
+            );
+            assert!(matches!(first_attempt, Ok(..)));
```
Comment (jstuczyn) [inline contracts/performance/src/transactions.rs]:
```suggestion
            assert!(first_attempt.is_ok())`
```

File: contracts/performance/src/storage.rs [hunk 7]
```diff
@@ -482,10 +542,12 @@ impl PerformanceResultsStorage {
         epoch_id: EpochId,
         data: &NodePerformance,
     ) -> Result<(), NymPerformanceContractError> {
+        self.assert_measurement_defined(storage, data.measurement_kind.clone())?;
```
Comment (jstuczyn) [inline contracts/performance/src/storage.rs]: this function should only really be concerned about writing data to storage and not checking whether it's actually allowed to insert it (i.e. ensuring kind exists). imo that should be done by the caller

File: contracts/performance/src/storage.rs [hunk 8]
```diff
@@ -192,12 +194,21 @@ impl NymPerformanceContractStorage {
 
         let mut accepted_scores = 0;
         let mut non_existent_nodes = Vec::new();
+        let mut non_existent_measurement_kind = Vec::new();
 
         // 3. submit it
         if self.node_bonded(deps.as_ref(), first.node_id)? {
```
Comment (jstuczyn) [inline contracts/performance/src/storage.rs:202]: this is just me thinking aloud. I wonder, would it perhaps simplify things if batch submission was always guaranteed to be for one kind at once?

File: contracts/performance/src/storage.rs [hunk 9]
```diff
@@ -482,10 +542,12 @@ impl PerformanceResultsStorage {
         epoch_id: EpochId,
         data: &NodePerformance,
     ) -> Result<(), NymPerformanceContractError> {
+        self.assert_measurement_defined(storage, data.measurement_kind.clone())?;
+
         let performance = data.performance;
 
-        let key = (epoch_id, data.node_id);
-        let updated = match self.results.may_load(storage, key)? {
+        let key = (epoch_id, data.node_id, data.measurement_kind.clone());
```
Comment (jstuczyn) [inline contracts/performance/src/storage.rs:555]: you could remove all the kind cloning all over the place if you changed your map key to `pub(crate) results: Map<(EpochId, NodeId, &'static str), NodeResults>,`

File: contracts/performance/src/queries.rs [hunk 10]
```diff
@@ -107,18 +140,19 @@ pub fn query_epoch_performance_paged(
         .unwrap_or(retrieval_limits::NODE_EPOCH_PERFORMANCE_DEFAULT_LIMIT)
         .min(retrieval_limits::NODE_EPOCH_PERFORMANCE_MAX_LIMIT) as usize;
 
-    let start = start_after.map(Bound::exclusive);
+    let start = start_after.map(|node_id| Bound::exclusive((node_id + 1, String::new())));
```
Comment (jstuczyn) [inline contracts/performance/src/queries.rs:143]: 1) wait, why are we skipping one node?
2) if you have to provide a rather opaque value to the bound, i.e. `String::new()`, perhaps the whole query signature should be changed? look at queries in other contracts, for pagination the caller is responsible for providing all elements of the subkey

File: contracts/performance/src/queries.rs [hunk 11]
```diff
@@ -142,27 +176,51 @@ pub fn query_epoch_measurements_paged(
         .unwrap_or(retrieval_limits::NODE_EPOCH_MEASUREMENTS_DEFAULT_LIMIT)
         .min(retrieval_limits::NODE_EPOCH_MEASUREMENTS_MAX_LIMIT) as usize;
 
-    let start = start_after.map(Bound::exclusive);
+    let start = start_after.map(|node_id| Bound::exclusive((node_id + 1, String::new())));
 
+    // because API aggregates per NodeId, and the storage doesn't, we have to
+    // first collect all different measurements for a node and use an
+    // intermediary struct to map from storage to the object returned on the API
+    let mut measurements_per_node: HashMap<NodeId, Vec<(MeasurementKind, NodeResults)>> =
```
Comment (jstuczyn) [inline contracts/performance/src/queries.rs:184]: I think we're just overcomplicating it here. we should just read the data as it's in the storage and let the caller sort it themselves

File: contracts/performance/src/queries.rs [hunk 12]
```diff
@@ -176,26 +234,63 @@ pub fn query_full_historical_performance_paged(
...
+    // so we need an intermediary struct that collects all measurements
+    // per (epoch_id, node_id) pair (and calculates a performance)
+    let mut res_per_epoch_and_node: HashMap<(EpochId, NodeId), HashMap<MeasurementKind, Percent>> =
+        HashMap::new();
+    NYM_PERFORMANCE_CONTRACT_STORAGE
         .performance_results
         .results
         .range(deps.storage, start, None, Order::Ascending)
-        .take(limit)
+        // we can't cut a pagination limit here becasue we don't want to
```
Comment (jstuczyn) [inline contracts/performance/src/queries.rs:251]: why not. just change the query parameters and let the caller resume from where it's meant to resume from

File: contracts/performance/src/storage.rs (hunk 9 above)
Comment (dynco-nym) [reply contracts/performance/src/storage.rs:555]: I don't think you can do that: this string is user-defined at runtime, you can't store it into `'static str`

File: contracts/performance/src/queries.rs (hunk 10 above)
Comment (dynco-nym) [reply contracts/performance/src/queries.rs:143]: `String::new()` is intended to start at `""` onwards for `MeasurementKind`. I.e. it's supposed to capture all measurement kinds under a certain node & epoch. 

In other pagination cases in the contract, you can "cut" anywhere across nodes & epochs. This signature guarantees that you cannot cut across measurement kinds of a node.

Comment (dynco-nym) [reply contracts/performance/src/queries.rs:143]: I don't see how this would be more ergonomic to paginate through

page 1
```
(epoch_1, node_1, measurement_1)
(epoch_1, node_1, measurement_2)
(epoch_1, node_1, measurement_3)
(epoch_1, node_2, measurement_1)
(epoch_1, node_2, measurement_2)
```

page 2 
```
(epoch_1, node_2, measurement_3)
(epoch_1, node_3, measurement_1)
(epoch_1, node_3, measurement_2)
(epoch_1, node_3, measurement_3)
(epoch_1, node_4, measurement_1)
```

but I'm open to it if you think it's more valuable

File: contracts/performance/src/queries.rs (hunk 11 above)
Comment (dynco-nym) [reply contracts/performance/src/queries.rs:184]: storage layout is a bit clunky due to cw-storage constraints. Don't we care about API being nicer?

File: contracts/performance/src/storage.rs (hunk 2 above)
Comment (dynco-nym) [reply contracts/performance/src/storage.rs]: done

File: contracts/performance/src/transactions.rs (hunk 4 above)
Comment (dynco-nym) [reply contracts/performance/src/transactions.rs]: done

File: contracts/performance/src/transactions.rs (hunk 6 above)
Comment (dynco-nym) [reply contracts/performance/src/transactions.rs]: fixed in several places

File: contracts/performance/src/storage.rs (hunk 9 above)
Comment (jstuczyn) [reply contracts/performance/src/storage.rs:555]: you totally can thanks to how cosmwasm contracts work : )
look at for example the `NetworkMonitorsStorage`. internally it uses `pub(crate) authorised: Map<&'static Addr, NetworkMonitorDetails>,`. those keys are also defined at runtime as new entries are being added

File: contracts/performance/src/queries.rs (hunk 10 above)
Comment (jstuczyn) [reply contracts/performance/src/queries.rs:143]: > In other pagination cases in the contract, you can "cut" anywhere across nodes & epochs. This signature guarantees that you cannot cut across measurement kinds of a node.

I think it's perfectly fine to allow cutting across measurements kinds of a node. after all this is a paged query to get all measurements of a node and multiple calls are expected

File: contracts/performance/src/queries.rs (hunk 11 above)
Comment (jstuczyn) [reply contracts/performance/src/queries.rs:184]: what I meant (I think, it's been 3 weeks : D) is that we shouldn't bother pre-sorting values in the response (i.e. assigning it to per node). instead we should just return raw rows as they are in the storage and leave it to the caller to deal with it like we do with all other queries
